src/libraries/System.Security.Cryptography/tests/KmacTestDriver.cs

```diff
@@ -0,0 +1,1024 @@
+            ReadOnlyMemory<byte> key,
+            Stream source,
+            int outputLength,
+            ReadOnlyMemory<byte> customizationString,
+            CancellationToken cancellationToken);
+
+        static abstract ValueTask<byte[]> HashDataAsync(
+            byte[] key,
+            Stream source,
+            int outputLength,
+            byte[] customizationString,
+            CancellationToken cancellationToken);
+    }
+
+    public abstract class KmacTestDriver<TKmacTrait, TKmac>
+        where TKmacTrait : IKmacTrait<TKmac>
+        where TKmac : IDisposable
+    {
+        // Test vectors can be generated with the following shell script. Change the algorithm and xof variables as needed.
+        //
+        // #!/usr/bin/env zsh
+        //
+        // algorithm="KMAC128" #KMAC128 or KMAC256
+        // xof="0" #0 or 1
+
+        // for keylen in 4 16 32
+        // do
+        //     key=$(seq 1 $keylen | awk '{ printf "%02x", $1 }')
+        //
+        //     for size in 1 {8..264..16}
+        //     do
+        //         for cs in 0 8
+        //         do
+        //             if [ $cs -eq 0 ]; then
+        //                 custom=""
+        //             else
+        //                 custom=$(seq 1 $cs | awk '{ printf "%02x", $1 }')
+        //             fi
+        //
+        //             for msg in "" "hello" "goodbye"
+        //             do
+        //                 msghex=$(echo -n $msg | xxd -p)
+        //                 mac=$(echo -e -n $msg | openssl mac -macopt hexkey:$key -macopt size:$size -macopt hexcustom:$custom -macopt xof:$xof $algorithm | awk '{print tolower($0)}')
+        //                 echo -n "yield return new(Key: \"$key\", Msg: \"$msghex\", Custom: \"$custom\", Mac: \"$mac\");\n"
+        //             done
+        //         done
+        //     done
+        // done
+        protected abstract IEnumerable<KmacTestVector> TestVectors { get; }
+
+        public static bool IsSupported => TKmacTrait.IsSupported;
+        public static bool IsNotSupported => !IsSupported;
+        public static KeySizes? PlatformKeySizeRequirements { get; } =
+            PlatformDetection.IsOpenSslSupported ? new KeySizes(4, 512, 1) : null;
+
+        public static int? PlatformMaxOutputSize { get; } = PlatformDetection.IsOpenSslSupported ? 0xFFFFFF / 8 : null;
+        public static int? PlatformMaxCustomizationStringSize { get; } = PlatformDetection.IsOpenSslSupported ? 512 : null;
+
+        public static byte[] MinimalKey { get; } =
+            PlatformKeySizeRequirements?.MinSize is int min ? new byte[min] : Array.Empty<byte>();
+
+
+        [ConditionalFact(nameof(IsSupported))]
+        public void KnownAnswerTests_Allocated_AllAtOnce()
+        {
+            foreach (KmacTestVector testVector in TestVectors)
+            {
+                using (TKmac kmac = TKmacTrait.Create(testVector.KeyBytes, testVector.CustomBytes))
```

Comment:
I don't know if it should be on all the KATs, or something else special, but there should be tests for null/empty equivalency on the customization string.

Reply:
There are KATs that explicitly pass in a null customization string. e.g.

https://github.com/dotnet/runtime/blob/60cc05be029f0a2dbc99e1b37925e5729bf277cf/src/libraries/System.Security.Cryptography/tests/Kmac128Tests.cs#L130

Or did you mean using the overload that omits passing the CS entirely?

Reply:
> null/empty equivalency on the customization string.

Ah okay. Re-read it this morning with my parser correctly caffeinated. You want a test proving that null is the same is empty.